### PR TITLE
Fix unstable with-items formatting

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

Fixes https://github.com/astral-sh/ruff/issues/10267

The issue with the current formatting is that the formatter flips between the `SingleParenthesizedContextManager`  and `ParenthesizeIfExpands` or `SingleWithTarget` because the layouts use incompatible formatting ( `SingleParenthesizedContextManager`: `maybe_parenthesize_expression(context)` vs `ParenthesizeIfExpands`: `parenthesize_if_expands(item)`, `SingleWithTarget`: `optional_parentheses(item)`. 

The fix is to ensure that the layouts between which the formatter flips when adding or removing parentheses are the same. I do this by introducing a new `SingleWithoutTarget` layout that uses the same formatting as `SingleParenthesizedContextManager` if it has no target and prefer `SingleWithoutTarget` over using `ParenthesizeIfExpands` or `SingleWithTarget`. 

## Formatting change

The downside is that we now use `maybe_parenthesize_expression` over `parenthesize_if_expands` for expressions where `can_omit_optional_parentheses` returns `false`. This can lead to stable formatting changes. I only found one formatting change in our ecosystem check and, unfortunately, this is necessary to fix the instability (and instability fixes are okay to have as part of minor changes according to our versioning policy)

The benefit of the change is that `with` items with a single context manager and without a target are now formatted identically to how the same expression would be formatted in other clause headers.

## Test Plan

I ran the ecosystem check locally

### CODE REVIEW

**Comment by aneeshusa** [comment]
I'm traveling today but am happy to give this a try tomorrow! Amazed how quick you put this together, thank you!

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/with.py [hunk 1]
```diff
@@ -13,45 +13,40 @@
     pass
     # trailing
 
-
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/with.py:16]: TODO: revert formatting changes

File: crates/ruff_python_formatter/src/other/arguments.rs [hunk 2]
```diff
@@ -205,11 +205,7 @@ fn is_arguments_huggable(arguments: &Arguments, context: &PyFormatContext) -> bo
 
     // If the expression has a trailing comma, then we can't hug it.
     if options.magic_trailing_comma().is_respect()
-        && commas::has_magic_trailing_comma(
-            TextRange::new(arg.end(), arguments.end()),
-            options,
-            context,
-        )
+        && commas::has_magic_trailing_comma(TextRange::new(arg.end(), arguments.end()), context)
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/other/arguments.rs:208]: I noticed that we pass `context` and `options`. Passing the `context` alone is sufficient because the options can be accessed by `context.options()`.

File: crates/ruff_python_formatter/src/statement/stmt_with.rs [hunk 3]
```diff
@@ -130,92 +133,182 @@ impl FormatNodeRule<StmtWith> for FormatStmtWith {
...
-fn should_parenthesize(
-    with: &StmtWith,
-    options: &PyFormatOptions,
-    context: &PyFormatContext,
-) -> FormatResult<ParenthesizeWith> {
-    if has_magic_trailing_comma(with, options, context) {
-        return Ok(ParenthesizeWith::IfExpands);
-    }
+#[derive(Clone, Copy, Debug)]
+enum WithItemsLayout<'a> {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_with.rs:137]: TODO: It might be worth to split the refactor into its own PR

**Comment by charliermarsh** [comment]
Thanks @aneeshusa :wave:

**Review by aneeshusa — COMMENTED** [review]
Can't do a GitHub approval but tested this and this fixes both an isolated repro and the monorepo at $WORK! Would love to see this in a release.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
Just shipped this in v0.3.2 :)
